### PR TITLE
Rebuild caches command

### PR DESCRIPTION
**Pull request by cviebrock**
As discussed in #33, this adds the `artisan eloquence:rebuild` command to rebuild one or more classes that implement a caching behaviour (i.e. `Cacheable` or `Summable`).

### CODE REVIEW

**Comment by cviebrock** [comment]
I haven't added any tests for the command ... partly because I'm not sure how, and partly because it looked like you were in the middle of refactoring the test suite.  Mind if I leave that work to you? ;)

File: src/Behaviours/Cacheable.php [hunk 1]
```diff
@@ -32,6 +32,30 @@ public function updateCacheRecord(array $config, $operation, $amount, $foreignKe
...
     }
 
+    public function rebuildCacheRecord(array $config, Model $model, $command, $aggregateField = null)
+    {
+
+        $table = $this->getModelTable($config['model']);
+        $modelTable = $this->getModelTable($model);
+
+        // the following is required for camel-cased models, in case users are defining their attributes as camelCase
+        $field = snake_case($config['field']);
```
Comment (kirkbushell) [inline src/Behaviours/Cacheable.php]: This behaviour is now pretty common. Any chance we can refactor the snake casing of these keys?

File: src/Behaviours/Cacheable.php [hunk 2]
```diff
@@ -32,6 +32,30 @@ public function updateCacheRecord(array $config, $operation, $amount, $foreignKe
...
+        $key = snake_case($this->key($config['key']));
+        $foreignKey = snake_case($this->key($config['foreignKey']));
+
+        if (is_null($aggregateField)) {
+            $aggregateField = $foreignKey;
+        } else {
+            $aggregateField = snake_case($aggregateField);
+        }
+
+        $sql = "UPDATE `{$table}` INNER JOIN (" .
```
Comment (kirkbushell) [inline src/Behaviours/Cacheable.php]: Nice. This is cool - will this work on all SQL implementations/drivers?

File: src/Commands/RebuildCaches.php [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+    /**
+     * Iterate through the application's classes and return all the ones
+     * that implement one of the caching behaviours.
+     *
+     * @return array[string]
+     */
+    private function getAllCacheableClasses()
+    {
+        $finder = new Finder;
```
Comment (kirkbushell) [inline src/Commands/RebuildCaches.php]: I'd extract this logic out into a separate class, and test that class separately, injecting the dependencies you need. That way you can test your logic appropriately.

File: src/Commands/RebuildCaches.php [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    /**
+     * Rebuilds the caches for the given class.
+     *
+     * @param string $className
+     */
+    private function rebuild($className)
+    {
+        $instance = new $className;
+
+        if (method_exists($instance, 'countCaches')) {
```
Comment (kirkbushell) [inline src/Commands/RebuildCaches.php]: Maybe do trait checks instead of method existence checks?

Comment (cviebrock) [inline src/Commands/RebuildCaches.php]: Sadly, I tried. But checking for traits only checks the current class, not any parent classes. So, while unlikely, your model might extend a base class that uses Cacheable, so you need to do a bunch of recursion.

Comment (kirkbushell) [inline src/Commands/RebuildCaches.php]: Oh fair enough, what a pain. hehe

Comment (kirkbushell) [inline src/Commands/RebuildCaches.php]: Even so, can we extract this out to its own service, maybe using a strategy pattern as well? That way if we implement further cache solutions (which is certainly in the pipeline), we'll be adhering to the O principle in SOLID, and allow for easy extension.

File: src/Behaviours/Cacheable.php (hunk 1 above)
Comment (cviebrock) [inline src/Behaviours/Cacheable.php]: Not sure how you suggest doing that.  A method that takes in a variable number of arguments by ref, and snake_cases them all?

Comment (kirkbushell) [inline src/Behaviours/Cacheable.php]: I'm not sure either. haha

Comment (cviebrock) [inline src/Behaviours/Cacheable.php]: I got something ... give me a sec

File: src/Behaviours/Cacheable.php (hunk 2 above)
Comment (cviebrock) [inline src/Behaviours/Cacheable.php]: Should do, yup

File: src/Commands/RebuildCaches.php (hunk 3 above)
Comment (cviebrock) [inline src/Commands/RebuildCaches.php]: Doing it ...

**Comment by cviebrock** [comment]
Have a look at some of the refactoring. I'm not sure I can spend a lot more time on this today, but feel free to take over if you think it's mostly there.

**Comment by kirkbushell** [comment]
Thanks @cviebrock.

I'll merge this in and work on it today, looks good - thanks for all your help! :)

**Comment by cviebrock** [comment]
No problem, thank you!  I suppose you're going to do a bit more cleanup before tagging another release?

**Comment by kirkbushell** [comment]
Plus test.etc. haven't tested the feature myself yet.

**Pull request merged**